src/module.c

```diff
@@ -485,71 +485,96 @@ void handleCompaction(RedisModuleCtx *ctx, CompactionRule *rule, api_timestamp_t
     RedisModule_CloseKey(key);
 }
 
+static inline int add(RedisModuleCtx *ctx, RedisModuleString *keyName, RedisModuleString *timestampStr, RedisModuleString *valueStr, RedisModuleString **argv, int argc){
+	  RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ|REDISMODULE_WRITE);
+
+	    double value;
+	    api_timestamp_t timestamp;
+	    if ((RedisModule_StringToDouble(valueStr, &value) != REDISMODULE_OK))
+	        return RedisModule_ReplyWithError(ctx, "TSDB: invalid value");
+
+	    if ((RedisModule_StringToLongLong(timestampStr, (long long int *) &timestamp) != REDISMODULE_OK)) {
+	        // if timestamp is "*", take current time (automatic timestamp)
+	        if(RMUtil_StringEqualsC(timestampStr, "*"))
+	            timestamp = (u_int64_t) time(NULL);
+	        else
+	            return RedisModule_ReplyWithError(ctx, "TSDB: invalid timestamp");
+	    }
+
+	    Series *series = NULL;
+
+	    if (argv != NULL && RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+	        // the key doesn't exist, lets check we have enough information to create one
+	        long long retentionSecs;
+	        long long maxSamplesPerChunk;
+	        size_t labelsCount;
+	        Label *labels;
+	        if (parseCreateArgs(ctx, argv, argc, &retentionSecs, &maxSamplesPerChunk, &labelsCount, &labels) != REDISMODULE_OK) {
```

Comment:
@gkorland there's an unexpected outcome for reusing this code. technically you made madd also to create the keys if they don't exists, not a problem really but it opens the door for an interesting use-case.

Reply:
madd always calls add with argv == NULL